File: ios/calorietracker/Services/HostedAIService.swift
//
//  HostedAIService.swift
//  calorietracker
//

import Foundation
import RevenueCat

enum HostedAIServiceError: LocalizedError {
    case invalidURL
    case unauthorized
    case serverError(String)
    case invalidResponse

    var errorDescription: String? {
        switch self {
        case .invalidURL: return "Invalid hosted AI URL."
        case .unauthorized: return "Hosted AI authentication failed."
        case .serverError(let msg): return msg
        case .invalidResponse: return "Could not understand the hosted AI response."
        }
    }
}

enum HostedAIService {
    private static let session: URLSession = {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 120
        return URLSession(configuration: config)
    }()

    static func generate(prompt: String, imageDataList: [Data], systemInstruction: String?) async throws -> String {
        let cappedImages = Array(imageDataList.prefix(HostedAIConstants.maxHostedImages))
        let body: [String: Any] = [
            "prompt": prompt,
            "images": cappedImages.map { $0.base64EncodedString() },
            "systemInstruction": systemInstruction as Any,
        ].compactMapValues { $0 }

        let data = try await post(path: "generate", jsonBody: body)
        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let text = json["text"] as? String else {
            throw HostedAIServiceError.invalidResponse
        }
        return text
    }

    static func geminiGenerate(requestBody: [String: Any]) async throws -> Data {
        try await post(path: "gemini", jsonBody: ["requestBody": requestBody])
    }

    static func transcribe(audioData: Data, mimeType: String, language: String?) async throws -> String {
        var body: [String: Any] = [
            "audio": audioData.base64EncodedString(),
            "mimeType": mimeType,
        ]
        if let language, !language.isEmpty {
            body["language"] = language
        }
        let data = try await post(path: "transcribe", jsonBody: body)
        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let text = json["text"] as? String else {
            throw HostedAIServiceError.invalidResponse
        }
        return text
    }

    private static func post(path: String, jsonBody: [String: Any]) async throws -> Data {
        guard let url = URL(string: "\(HostedAIConstants.hostedAIBaseURL)/\(path)") else {
            throw HostedAIServiceError.invalidURL
        }

        let userID = await RevenueCatManager.shared.appUserID()
        let plan = await RevenueCatManager.shared.activePlan.rawValue

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(HostedAIConstants.hostedAIAppSecret)", forHTTPHeaderField: "Authorization")
        request.setValue(userID, forHTTPHeaderField: "X-Fud-User-Id")
        request.setValue(plan, forHTTPHeaderField: "X-Fud-Plan")
        request.httpBody = try JSONSerialization.data(withJSONObject: jsonBody)

        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw HostedAIServiceError.invalidResponse
        }
        if http.statusCode == 401 {
            throw HostedAIServiceError.unauthorized
        }
        if !(200...299).contains(http.statusCode) {
            let message = (try? JSONSerialization.jsonObject(with: data) as? [String: Any])?["error"] as? String
            throw HostedAIServiceError.serverError(message ?? "Hosted AI request failed (\(http.statusCode)).")
        }
        return data
    }
}

File: android/app/src/main/java/com/apoorvdarshan/calorietracker/services/ai/HostedAIService.kt
package com.apoorvdarshan.calorietracker.services.ai

import com.apoorvdarshan.calorietracker.billing.HostedAIConstants
import com.apoorvdarshan.calorietracker.billing.HostedPlan
import com.apoorvdarshan.calorietracker.billing.RevenueCatManager
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put
import kotlinx.serialization.json.putJsonArray
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import java.util.concurrent.TimeUnit

class HostedAIService(
    private val client: OkHttpClient = defaultClient,
    private val revenueCat: RevenueCatManager
) {
    suspend fun generate(
        prompt: String,
        imageBytesList: List<ByteArray>,
        systemInstruction: String?
    ): String = withContext(Dispatchers.IO) {
        val images = buildJsonObject {
            putJsonArray("images") {
                imageBytesList.take(HostedAIConstants.MAX_HOSTED_IMAGES).forEach { bytes ->
                    add(kotlinx.serialization.json.JsonPrimitive(android.util.Base64.encodeToString(bytes, android.util.Base64.NO_WRAP)))
                }
            }
            put("prompt", prompt)
            systemInstruction?.takeIf { it.isNotBlank() }?.let { put("systemInstruction", it) }
        }
        val data = post("generate", images.toString())
        JSONObject(data).getString("text")
    }

    suspend fun geminiGenerate(requestBody: JSONObject): JSONObject = withContext(Dispatchers.IO) {
        val wrapper = JSONObject().put("requestBody", requestBody)
        JSONObject(post("gemini", wrapper.toString()))
    }

    suspend fun transcribe(audioBytes: ByteArray, mimeType: String, language: String?): String =
        withContext(Dispatchers.IO) {
            val body = buildJsonObject {
                put("audio", android.util.Base64.encodeToString(audioBytes, android.util.Base64.NO_WRAP))
                put("mimeType", mimeType)
                language?.takeIf { it.isNotBlank() }?.let { put("language", it) }
            }
            val data = post("transcribe", body.toString())
            JSONObject(data).getString("text")
        }

    private suspend fun post(path: String, jsonBody: String): String {
        val plan = revenueCat.activePlan.value
        val userId = revenueCat.appUserId()
        val request = Request.Builder()
            .url("${HostedAIConstants.HOSTED_AI_BASE_URL}/$path")
            .post(jsonBody.toRequestBody(JSON_MEDIA))
            .header("Authorization", "Bearer ${HostedAIConstants.HOSTED_AI_APP_SECRET}")
            .header("X-Fud-User-Id", userId)
            .header("X-Fud-Plan", plan.id)
            .build()
        client.newCall(request).execute().use { response ->
            val body = response.body?.string().orEmpty()
            if (!response.isSuccessful) {
                val detail = runCatching { JSONObject(body).optString("error") }.getOrNull()
                throw AiError.Api(detail ?: "Hosted AI failed (${response.code})")
            }
            return body
        }
    }

    companion object {
        private val JSON_MEDIA = "application/json".toMediaType()
        private val defaultClient = OkHttpClient.Builder()
            .connectTimeout(120, TimeUnit.SECONDS)
            .readTimeout(120, TimeUnit.SECONDS)
            .writeTimeout(120, TimeUnit.SECONDS)
            .build()
    }
}
